### ISSUE TITLE
UX: Vertical toolbar size and controls grouping

### ISSUE THREAD

**Issue by oleq** [ckeditor/ckeditor5-design#95]
# How to present a rich toolbar using minimum space?
- Which kind of UI should be preferred and promoted by us as default (like a good example)?
  - Single toolbar with **top menus**, **tabs** or **button dropdowns**?
    - How does it affect a11y?
  - Multi–row toolbar (like in CKEditor 3+)?
  - **Note:** This point is very, very important because developers will follow those examples and **this will eventually become how the outside world perceives CKE5** (heavy and cluttered vs lightweight and smart), it's also very hard to undo this first impression, no matter how many different UI implementations CKE5 would offer in the future.
- If multi–row toolbars are _a thing of the past_, then which is better for overall UX: **top menus**, **tabbed toolbar** or **button dropdowns**?
  - Can any of those solutions be safely combined (i.e. tabbed toolbar + button dropdowns)?
- What other solutions can be offered?

---
## Multi-line toolbar

![image](https://cloud.githubusercontent.com/assets/1099479/12417122/28c3c098-bea6-11e5-8bf7-6fd649ca1aad.png)
## A single-line toolbar with as much grouping as possible

![image](https://cloud.githubusercontent.com/assets/1099479/12417132/34689e00-bea6-11e5-8bc1-3fbe2f9bdce3.png)
## A global top menu with dropdowns

![image](https://cloud.githubusercontent.com/assets/1099479/12417138/399a3118-bea6-11e5-866b-e4913e2c52bd.png)

![image](https://cloud.githubusercontent.com/assets/1099479/12417141/3e028426-bea6-11e5-956a-ae9451b5fd8a.png)
## A tabbed toolbar (multiple toolbars)

![image](https://cloud.githubusercontent.com/assets/1099479/12417151/42d28e92-bea6-11e5-8564-25279b2c3729.png)

![image](https://cloud.githubusercontent.com/assets/1099479/12417153/460afb26-bea6-11e5-8876-0f75aec8cc3f.png)

**Comment by fredck** [comment]
Do we need "select boxes / combos" at all? Should we also discuss if they should be represented in a different way (e.g. a "select button"). Maybe a dedicated ticket should be created for that but wanted to comment it here because it impacts drastically on the distribution of elements inside the toolbar.

In any case, I'm sure that some people would still like to have select boxes anyway. I think we're mainly talking about the default implementation here.

**Comment by oleq** [comment]
> Should we also discuss if they should be represented in a different way (e.g. a "select button").
By "select button" you mean a standard button which, when clicked, shows a dropdown?

**Comment by fredck** [comment]
> By "select button" you mean a standard button which, when clicked, shows a dropdown?

Yes. Imagine the very same "select box" behavior but activated by a button. You click on it and the selection list dropdown appears.

**Comment by oleq** [comment]
> Yes. Imagine the very same "select box" behavior but activated by a button. You click on it and the selection list dropdown appears.

But you lose the preview this way. I'll create a separate ticket for discussion in a minute.

**Comment by oleq** [comment]
@fredck Please let me know if #101 reflects your concerns.

**Comment by fredck** [comment]
> @fredck Please let me know if #101 reflects your concerns.

Yes, that's exactly what I was talking about. Thanks!

**Comment by wimleers** [comment]
>  this will eventually become how the outside world perceives CKE5

+1

I think @koleary will have thoughts on this, as well as https://twitter.com/bojhan.

**Comment by smartcorestudio** [comment]
+1 for ability to have button dropdowns, even in multi-line toolbar. Dropdowns can look like classic menus with icons. For example it can be used for align, paste, insert options etc.

**Comment by ofcapl** [comment]
+1 for classic dropdowns & buttons, grouped in tabs. Keep it simple, keep it classy. It works since first M$ Words era so people get used to it and it makes them productive while using such tools.

**Comment by SteveTheTechie** [comment]
Ok, since I posted about this topic a long time ago... here are are my thoughts:

1.  As I have mentioned in that past post, a key consideration is what functionality needs to be *visible* vs what needs to be *available* on a toolbar or menu.   The current 4.x toolbar tends to make everything visible, which contributes to clutter and user info overload.  Drop-downs and menus allow things to be available while not necessarily being visible.

2.  Because of #1 above, I am very much in favor of button drop-downs.  +1 on that.  They allow you to hide some things while keeping them available.

3.  I prefer drop downs and menus, but other people may prefer tool bars.  (I do like toolbars for frequently used commands.)  Flexibility is needed.  For example, in the current CKE 4.x UI, it recently occurred to me that a useful plug-in would have been to show a horizontal bar menu when the toolbar is in collapsed state.  This would take care of the "toolbar crowd" and the "menu crowd".  Of course, it is really too late for such a plug-in for 4.x, but keep in mind that some similar type of flexibility would probably be helpful.  Perhaps you have a menu div followed by a toolbar div in the markup... and one or both could be used and populated depending on what configuration of the UI is being used.

4.  I think a grouping mechanism for toolbar controls and menu entries would help.  Groups/sub-groups in toolbars allow for the tabbed implementation (could be via a plug-in).  The tabbed implementation could use the "menu div" from #3 above for the tab names.  Groups/sub-groups in menus allow for sub-menuing.

**Comment by fredck** [comment]
Let's also keep in mind that there are two kinds of expectations when it comes to the editor UI. (1) One more cluttered where the editor UI take an important role, like an application UI, and (2) a new tendency that we see where the editor UI is pretty minimalistic, to not interfere with the with the whole application UI. I think we should try to figure out how to make "2" happen because it seems to be more broadly acceptable.

**Comment by koleary** [comment]
Ok I'm going to break out of the box a little here. 

I think this whole discussion centers around the use of space and how you can more efficiently get stuff to fit and that is somewhat missing the point. What is most important to the user is that they are presented with the _specific_ tool that they need at the moment, and in the place, that they need it.

Doing this is not a simple task. Let's say for example Ann is a user on a site where the default toolbar has been configured to have 20 buttons and 12 drop-down options in two groups, or a total of 30 options. Ann works in the legal department and as such the button she most commonly uses is strikethrough, in her default editor configuration the 17th button in the toolbar.

To "improve" the authoring experience the admin team decides to go with an inline toolbar, but the inline toolbar needs to have fewer visible buttons because the boss said "I want it to be simple, like Medium". Now Ann's strikethrough button is the tenth of 11 in a dropdown and her experience is painful because every time she uses the button she has to click the dropdown and move across 9 small buttons to get strikethrough and if her hand moves a bit off the dropdown, it collapses and she has to try again.

The ideal experience for Ann would be that she has a compact, inline bar with three or four options and one of them is strikethrough. In short the ideal experience for Ann is an editor that shows her the tools _she_ needs,  the tools _everyone_ needs. 

So how do we get there? In the past applications have enabled users to set per-user preferences and desktop applications often "remember" the last panel or tool you used or where you dragged it to. Neither of those are really effective solutions because Ann doesn't want to have to "set" her preferences and the last tool she used is not necessarily the one she needs now.

What Ann needs is a system that automatically remembers _the tools she uses most commonly_ and presents only those to her, with some way of accessing the other available tools if she needs them. 

I am a designer, not a developer so I'm not going to tell you how to build it. Obviously for CMSs this information could be in the user table, but I imagine that this could be made to work for anonymous users as well with some combination of persistent cookies and local storage.

From a UX perspective what you need is an in place toolbar with a limited number of buttons and an "expand" toggle that shows the rest. Use of a button by a (cookied) user would simply increment up the weight of the button and the heaviest buttons float upwards. They may never significantly change the default weighting, for example if** B _I_ U **are the defaults and I most commonly use **B**, but they will dramatically simplify the overall experience and improve it for the outliers.

**Comment by koleary** [comment]
Also, just to clarify, when I say in-place editing that is not confined to an end-user facing page. The "floating" toolbar could and should also be used in the context of creating or editing content that is not published (as it is in Medium).

**Comment by Reinmar** [comment]
Thanks @koleary. Very interesting point. When the space is limited (either because the developers made such a choice, or the device's screen is small) such a feature could significantly improve the UX.

I wonder though how often the buttons could change their order to not confuse the user. Can the buttons change their order "live" (as the user is editing some content) or should that happen only after page reload?

Another danger is when the user works in two or more "modes". One day he/she is writing some text and the next day he/she is correcting it. That could lead to always displaying a toolbar with the wrong buttons :D.

**Comment by koleary** [comment]
Good points. 

I'm thinking that the weighting problem could be handled by incrementing the weight by 1 and spreading the default weights of the buttons. So for example if the default button weights are in increments of 10 like so:

1. Bold: 50
2. Italic: 40
3. Underline: 30
4. Strikethrough: 20
5. Link: 10

Then if each time a user uses a button it gets an additional point of weight, it would need to be used 10 times before it changes position, assuming no other button is used at all. That way the re-ordering of the buttons would happen at a much slower and more subtle rate and would be less disruptive to the experience (a button wouldn't move more than one "step" based on a single action). Nevertheless the buttons would seem to "magically" adapt to the needs of the user.

Even in the case where you have different "modes" the method would still produce a set with the most common buttons for that user for both tasks, and in any event, it's unlikely that those sets would not intersect. 

I'm not a mathematician, and I'm certain there is probably a simpler way to handle an indirect relationship like this with floats or negative and positive integers or something, but I think you get the point which is that a button should be used several more times relative to the rate of use of the other buttons before it changes position.

**Comment by Reinmar** [comment]
Cleaning up outdated discussions. See https://github.com/ckeditor/ckeditor5-design/issues/186
